oshdb/src/main/java/org/heigit/ohsome/oshdb/util/bytearray/ByteArrayWrapper.java

```diff
@@ -62,163 +57,70 @@ public int hasLeft() {
-    return readRawVarint32();
+  public int readU32() {
+    int b = buffer[bufferPos++];
+    if (b >= 0) {
+      return b;
+    }
+
+    int value = b & 0x7F;
+    var i = 7;
+    while (((b = buffer[bufferPos++]) & 0x80) != 0) {
+      value |= (b & 0x7F) << i;
+      i += 7;
+    }
+    return value | b << i;
   }
 
   /**
    * Read an {@code sint64} field value from the stream.
    */
-  public long readS64() throws IOException {
-    return decodeZigZag64(readRawVarint64());
+  public long readS64() {
+    return decodeZigZag64(readU64());
   }
 
-  public long readS64Delta(long last) throws IOException {
+  public long readS64Delta(long last) {
     return readS64() + last;
   }
 
   /**
    * Read a {@code uint64} field value from the stream.
    */
-  public long readU64() throws IOException {
-    return readRawVarint64();
-  }
-
-  public long readU64Delta(long last) throws IOException {
-    return readU64() + last;
-  }
-
-  /**
-   * Read a raw Varint from the stream. If larger than 32 bits, discard the upper bits.
-   */
-  public int readRawVarint32() throws IOException {
-    // See implementation notes for readRawVarint64
-    fastpath: {
-      int pos = bufferPos;
-
-      if (bufferSize == pos) {
-        break fastpath;
-      }
-
-      final byte[] buffer = this.buffer;
-      int x;
-      if ((x = buffer[pos++]) >= 0) {
-        bufferPos = pos;
-        return x;
-      } else if (bufferSize - pos < 9) {
-        break fastpath;
-      } else if ((x ^= buffer[pos++] << 7) < 0L) {
-        x ^= ~0L << 7;
-      } else if ((x ^= buffer[pos++] << 14) >= 0L) {
-        x ^= ~0L << 7 ^ ~0L << 14;
-      } else if ((x ^= buffer[pos++] << 21) < 0L) {
-        x ^= ~0L << 7 ^ ~0L << 14 ^ ~0L << 21;
-      } else {
-        int y = buffer[pos++];
-        x ^= y << 28;
-        x ^= ~0L << 7 ^ ~0L << 14 ^ ~0L << 21 ^ ~0L << 28;
-        if (y < 0 && buffer[pos++] < 0 && buffer[pos++] < 0 && buffer[pos++] < 0
-            && buffer[pos++] < 0 && buffer[pos++] < 0) {
-          break fastpath; // Will throw malformedVarint()
-        }
-      }
-      bufferPos = pos;
-      return x;
+  public long readU64() {
+    long b = buffer[bufferPos++];
+    if (b >= 0) {
+      return b;
     }
-    return (int) readRawVarint64SlowPath();
-  }
-
-  /** Read a raw Varint from the stream. */
-  public long readRawVarint64() throws IOException {
-    // Implementation notes:
-    // Optimized for one-byte values, expected to be common.
-    // The particular code below was selected from various candidates
-    // empirically, by winning VarintBenchmark.
-    // Sign extension of (signed) Java bytes is usually a nuisance, but
-    // we exploit it here to more easily obtain the sign of bytes read.
-    // Instead of cleaning up the sign extension bits by masking eagerly,
-    // we delay until we find the final (positive) byte, when we clear all
-    // accumulated bits with one xor. We depend on javac to constant fold.
```

Comment:
is this "fast path" not needed anymore?

Reply:
Actually not, this implementation was originally taken from the google protobuf library. I also looked into other implementations like SerializationUtils from apache orc or BytesUtils from apache parquet both use a more simpler and more maintainable version of varint decoding. so I opt to do the optimization for one-byte values, like protobuf but then switch to the simpler implementation like orc or parquet.

Reply:
:+1: thanks for the explanation.